### PR TITLE
RW-11654 Added infrastructure necessary to support contract tests with runtimes.

### PR DESCRIPTION
**Pull request by evrii**
__Jira ticket__: [RW-11654](https://precisionmedicineinitiative.atlassian.net/browse/RW-11654)

### What

- Added a couple extra states to handle runtime functionality
- Separated states into area of concern

### Why

- With the potential of dozens of states being added, I believe that handling all of the state management in a single file would end up being cumbersome.

### What to test

- 

### Who tested and where

- [ ] This change is covered by automated tests 
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [ ] I validated this change 
- [ ] Primary reviewer validated this change 
- [ ] I validated this change in the dev environment

### CODE REVIEW

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (evrii) [inline pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:73]: Move state management functionality to separate files depending on the area of the app. This file only cares about combining the other state management functions.

**Comment by ichengchang** [comment]
You can update the following code in your PR

```plaintext
# pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala: 92 in eric/RW-11654
```
to

```
case _                     => consumerVersionSelectors = consumerVersionSelectors.deployedOrReleased.mainBranch.branch("eric/RW-11654", "aou-rwb-api")
```

to test your consumer contract.

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (ichengchang) [inline pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala]:
```suggestion
    case _                     => consumerVersionSelectors = consumerVersionSelectors.deployedOrReleased.branch("eric/RW-11654", Option("aou-rwb-api"))
```

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (evrii) [inline pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:95]: Do not merge this.

File: .github/workflows/verify_consumer_contracts.yml
Comment (ichengchang) [inline .github/workflows/verify_consumer_contracts.yml]: Removed this before merge

File: .github/workflows/verify_consumer_contracts.yml
Comment (ichengchang) [inline .github/workflows/verify_consumer_contracts.yml]: remove this before merge

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (ichengchang) [reply pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:95]: Remember to revert before merge

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (ichengchang) [inline pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:149]: can try this

```
.withPendingPactsEnabled(ProviderTags(providerVer))
```

and

```
.withConsumerVersionSelectors(consumerVersionSelectors)
```

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (jdcanas) [inline pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:74]: huh... a little confused why there are multiple `orElse` chained here, seems like a bit of a code smell but I'm not super familiar with what this code is doing enough to know for sure. I'd recommend trying to change this to a function that has a match statement if possible, psuedocode below
```
determiningThing match {
case 1 =>
case 2 =>
case _ => loggerIO.debug("Whoops: other state")
}
```

Comment (ichengchang) [reply pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:74]: I think chaining partial functions encourages separation of duties and re-usability of code (like a runtime state manager and another state manager for the rest of the Leo app), the effect is going to be the same as using a `case match` block.

Comment (jdcanas) [reply pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:74]: the downside is its not super clear why one is taking precedence over another (which is what my main point of confusion is). Don't feel super strong about switching the syntax though since semantically it would be the same, a comment could suffice

**Review by jdcanas — APPROVED** [review]
small scala related nit, otherwise looks good

File: pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala
Comment (ichengchang) [reply pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:74]: If the states are mutually exclusive then I think there is no precedence one way over the other. @evrii would you add a comment to your code for clarity sake?

Comment (evrii) [reply pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala:74]: Thanks @jdcanas and @ichengchang .

I added a comment for clarification.

@jdcanas I would prefer a case statement here too, but I have not been able to figure out a way to handle that.

```
  // This function composes a large switch statement based on partial functions from
  // multiple state mangers (each with their own state that they handle).
  private val providerStatesHandler: StateManagementFunction = StateManagementFunction {
    RuntimeStateManager.handler(mockRuntimeService).orElse(AppStateManager.handler(mockAppService)).orElse { case _ =>
      loggerIO.debug("Whoops: other state")
    }
  }
```

The way this works is that a given state is passed in (say X in this case).

This function checks RuntimeStateManager and it finds that it has cases for A,B, and C but no X

It then proceeds to AppStateManager where is finds states X, Y, and Z. So it executes the code in the X case.

I could see this working like this, but I haven't been able to get it to work:
```
case (A|B|C):
   Call RuntimeStateManager
      case A:
           ...
      case B:
           ...
      case C:
           ...
case (X|Y|Z):
    Call AppStateManager
      case X:
           ...
      case Y:
           ...
      case Z:
           ...
```

**Pull request merged**
